### PR TITLE
Add full type declarations for projects and reader support

### PR DESCRIPTION
**Pull request by programmersstudio**
The aim of the PR is to reduce the dependency on the deployer by moving the project reader/writer service here. The deployment part of the project is still being handled by the deployer.

This PR only contains the project reader service.

### CODE REVIEW

**Review by joshuaauerbachwatson — CHANGES_REQUESTED** [review]
Nice, overall, @programmersstudio.   Some required changes and a few comments that don't require action (hopefully clear which are which).   Also, please fill in the description saying briefly what this PR does.

Note to the `doctl` team: this code will be dormant for a while, but it is useful to us to have it merged and in the codebase going forward because many other planned changes will build on it.

File: commands/serverless-extra.go
Comment (joshuaauerbachwatson) [inline commands/serverless-extra.go:162]: Sorry.   My previous suggestion was slightly broken.   That's the problem with the github suggestion thing.  It encourages the  reviewer to propose (and the PR author to accept) untested code.

```suggestion
		ProjectPath: c.Args[0],
```

File: commands/serverless-extra.go
Comment (joshuaauerbachwatson) [inline commands/serverless-extra.go:165]:
```suggestion
		output, err = c.Serverless().ReadProject(&project, c.Args)
```

**Review by joshuaauerbachwatson — CHANGES_REQUESTED** [review]
Sorry.   A previous suggestion of mine was broken.   I suggested a fix.   This time I tested it :-).   So, we need one more tweak.

File: commands/serverless-extra.go
Comment (programmersstudio) [reply commands/serverless-extra.go]: Not a problem, I noticed that and fixed and tested it in the local system.

**Review by joshuaauerbachwatson — APPROVED** [review]
Looks good to me now.   Thanks, @programmersstudio .

 I'm going to wait for an approval from a member of the doctl team before merging.

We are going to do a feature merge and public release toward the end of the month and this code will be there (dormant).

File: commands/serverless-extra.go
Comment (andrewsomething) [inline commands/serverless-extra.go:65]: I'm curious about the purpose of this flag. Is only intended for internal testing, or will we expose it to customers? Think I'm just missing some context.

Comment (joshuaauerbachwatson) [reply commands/serverless-extra.go:65]: It will never be exposed to customers.    The plan is that `get-metadata` will _always_ using the new project-reader service once we've gained enough confidence to switch over.   The flag is "for testing only" while we're not yet ready to switch, and will help us gain the confidence by allowing us to compare behavior with and without the flag.   So the flag will eventually be removed without ever being exposed.

File: do/serverless.go
Comment (scotchneat) [inline do/serverless.go:695]: Even though this code path is hidden, since we're not doing anything with the `readProjectConfig` spec output here, it might make sense to return an error message like this until it's fully implemented (unless this is intended behavior):

```suggestion
	_, err = readProjectConfig(project.ConfigPath)
	if err != nil {
		return ServerlessOutput{}, err
	}
	return ServerlessOutput{}, fmt.Errorf("not implemented")
```

**Review by scotchneat — COMMENTED** [review]
Looks good overall. I added a couple of comments inline.

**Pull request merged**
